### PR TITLE
Add Drupal Dependencies commands to Drush core

### PR DESCRIPTION
**Pull request by claudiu-cristea**
Add https://github.com/claudiu-cristea/drupal-dependencies to Drupal core

### CODE REVIEW

**Comment by claudiu-cristea** [comment]
Blocked on #6061

**Comment by claudiu-cristea** [comment]
Unblocking. Resolving w/o #6061 for now

**Comment by claudiu-cristea** [comment]
Ok, let's see what Drush maintainers have to say.

In the future we can expand with more `why:*` commands, such as getting the container services dependencies, Drupal libraries dependencies, etc.

File: docs/drupal-dependencies.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+:octicons-tag-24: 13.4+
+
+These commands allow a developer or site builder to inspect the Drupal dependencies. It's similar with Composer's `why` command but acts in the Drupal realm, by showing dependencies between modules or config entities.
+
+Find module dependants
+----------------------
+
+Drupal modules are able to define other modules as dependencies, using the module's [metadata info.yml file](https://www.drupal.org/docs/develop/creating-modules/let-drupal-know-about-your-module-with-an-infoyml-file). To get all modules that depend on a given module type:
+
+    drush why:module node --dependent-type=module
```
Comment (weitzman) [inline docs/drupal-dependencies.md]: How about just --type? Its briefer, and easier to spell.

File: src/Commands/core/DrupalDependenciesCommands.php [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+    private const CIRCULAR_REFERENCE = '***circular***';
+    private array $dependents = [];
+    private array $tree = [];
+    private array $relation = [];
+    private array $dependencies = [
+      'module-module' => [],
+      'config-module' => [],
+      'config-config' => [],
+    ];
```
Comment (weitzman) [inline src/Commands/core/DrupalDependenciesCommands.php]: PHPDoc for these would be helpful IMO

File: src/Commands/core/DrupalDependenciesCommands.php [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            throw new \InvalidArgumentException(dt('Invalid @config config entity', [
+                '@config' => $configName,
+            ]));
+        }
+    }
+
+    /**
+     * @param string $dependency
+     * @param array $path
+     */
```
Comment (weitzman) [inline src/Commands/core/DrupalDependenciesCommands.php]: Lets do a 1 line summary of the method, and omit phpdoc on params if they add nothing above whats in the function signature. Same for methods below.

File: src/Commands/core/DrupalDependenciesCommands.php [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+    }
+
+    /**
+     * @param bool $scanModuleDependencies
+     */
+    protected function scanConfigs(bool $scanModuleDependencies = true): void
+    {
+        $entityTypeManager = \Drupal::entityTypeManager();
+        $configTypeIds = array_keys(
+            array_filter($entityTypeManager->getDefinitions(), function (EntityTypeInterface $entityType): bool {
```
Comment (weitzman) [inline src/Commands/core/DrupalDependenciesCommands.php]: could you improve readability with an arrow function here? https://medium.com/@albertcolom/how-to-use-arrow-function-in-php-c28490ff7fb7

**Review by weitzman — COMMENTED** [review]
Great job on the tests. I have just nits.

File: src/Commands/core/DrupalDependenciesCommands.php [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+        private readonly array $installedModules,
+    ) {
+        parent::__construct();
+    }
+
+    public static function create(ContainerInterface $container): self
+    {
+        return new self(
+            $container->get('extension.list.module'),
+            $container->getParameter('container.modules')
```
Comment (claudiu-cristea) [inline src/Commands/core/DrupalDependenciesCommands.php:70]: This container param I wanted to autowire, so I've opened #6061

**Comment by claudiu-cristea** [comment]
Ready for a 2nd round of review

**Comment by weitzman** [comment]
This looks great to me. I will merge in a day or two  unless we get more discussion here.

**Pull request merged**